Drive bugs

MrFlashAccount opened the pull request:
### Pull Request Description

Closes: https://github.com/enso-org/cloud-v2/issues/1493

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

MrFlashAccount committed:
Fix remove button tooltip in sidebar

MrFlashAccount committed:
Implementation of the Path column

MrFlashAccount committed:
Dispay group names properly

MrFlashAccount committed:
Fixes

MrFlashAccount committed:
Small fixes

MrFlashAccount committed:
Prettier

MrFlashAccount committed:
FIx tests

MrFlashAccount committed:
Fix lint

MrFlashAccount committed:
Improvements

MrFlashAccount committed:
Merge branch 'develop' into wip/sergeigarin/drive-bugs

# Conflicts:
#	app/gui/src/dashboard/layouts/Drive.tsx

MrFlashAccount committed:
Fix tests

MrFlashAccount committed:
fix prettier

MrFlashAccount committed:
Small fixes

MrFlashAccount committed:
Fix tests, small refactor

MrFlashAccount committed:
Revert Await component in the Drive

MrFlashAccount committed:
Fix eslint

MrFlashAccount committed:
Add tests, fix some bugs

MrFlashAccount committed:
Try fix unit tests

MrFlashAccount committed:
Fix types

MrFlashAccount committed:
Fix eslint

PabloBuchu commented:
1. Wen to trash -> clicked `Analytics` in location -> ended up with minor styling glitch
<img width="425" alt="Screenshot 2024-12-24 at 12 33 58" src="https://github.com/user-attachments/assets/b2675c3b-9e5a-4196-b527-8a3bc0611ad8" />

2. Not sure if the location should be clickable. Can we have full path on hover?
3. Remove from sidebar modal still says "remove local forlder from favourites"
<img width="547" alt="Screenshot 2024-12-24 at 12 36 19" src="https://github.com/user-attachments/assets/d61676b6-bf65-495e-b891-37cc3397cf48" />

4. Above modal `Remove` does nothing. The folder is still in the sidebar
5. `Recent` location looks much nicer than trash one. Providing more information. Can we unify them?
<img width="278" alt="Screenshot 2024-12-24 at 12 38 37" src="https://github.com/user-attachments/assets/49fa1534-121a-49bc-911d-cce300c31a91" />

MrFlashAccount committed:
Merge branch 'develop' into wip/sergeigarin/drive-bugs

# Conflicts:
#	app/gui/integration-test/dashboard/actions/api.ts
#	app/gui/integration-test/dashboard/assetsTableFeatures.spec.ts
#	app/gui/src/dashboard/App.tsx
#	app/gui/src/dashboard/components/dashboard/AssetRow.tsx
#	app/gui/src/dashboard/layouts/AssetsTable.tsx

MrFlashAccount commented:
> Wen to trash -> clicked Analytics in location -> ended up with minor styling glitch

FIxed ✅

> Not sure if the location should be clickable. Can we have full path on hover?

Hmmm... My first implementation was on hover. But it causes issues - when user just move the cursor across the table, the popup appears and cover the asset below. This was annoying. So I ended up with the click.

There's an inconsistent behavior though - if the asset located in the root, we don't show popup on click, instead we navigate to the folder. But I think it's distincted well enough

> Remove from sidebar modal still says "remove local folder from favourites"

Thanks, fixed ✅

> Above modal Remove does nothing. The folder is still in the sidebar

Fixed ✅

> Recent location looks much nicer than trash one. Providing more information. Can we unify them?

WDYM? They're exactly the same components

MrFlashAccount commented:
> Wen to trash -> clicked Analytics in location -> ended up with minor styling glitch

This is a tricky bug, not simple to fix. I need more time to investigate and do a proper fix

MrFlashAccount committed:
Address issues from QA

MrFlashAccount committed:
Fix prettier

MrFlashAccount committed:
Fix prettier x2

MrFlashAccount committed:
Fix navigation

MrFlashAccount committed:
Remove console.log

MrFlashAccount committed:
Fix headers

somebody1234 commented on app/gui/integration-test/dashboard/actions/DrivePageActions.ts:
```diff
@@ -229,6 +270,28 @@ export default class DrivePageActions<Context> extends PageActions<Context> {
+
+          await expect(expandButton).toHaveAttribute('aria-label', TEXT.expand)
+
+          return expandButton.click()
```
nit: i prefer `await`ing all promises even if the result is the same

MrFlashAccount replied:
Even though _techically_ there's no difference between `return` and `await`, we throw the return of the callback anyway, so yeah, await here is preferable.

somebody1234 commented on app/gui/src/dashboard/components/AriaComponents/Button/CloseButton.tsx:
```diff
@@ -28,14 +29,15 @@ export const CloseButton = memo(function CloseButton(props: CloseButtonProps) {
-          // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+          // @ts-expect-error ts fails to infer the type of the className prop
+          // But it's safe because we are passing all values transparently
+          // and they typed outside
```
`they` -> `they are`, `ts` -> `TypeScript`, end with full stop, `className` -> `` `className` ``

somebody1234 commented on app/gui/src/dashboard/utilities/tailwindVariants.ts:
```diff
@@ -22,10 +22,10 @@ export type ExtractFunction<T> =
-  'class' | 'className'
-> & {
-  variants?: ExtractFunction<T> | undefined
+// TODO: add support for styling individual slots
```
```
// TODO[sg]: link to issue here
// Add support for styling individual slots.
```

MrFlashAccount replied:
Added, though the format is a little bit different: 
`  * TODO: @MrFlashAccount [add support for styling individual slots](https://github.com/enso-org/cloud-v2/issues/1643) `

- Direct reference to an GH account
- Clickable link to the issue with readable title
- Support for highlighting in vscode with [TODO highlight](https://marketplace.visualstudio.com/items?itemName=wayou.vscode-todo-highlight)

Thoughs?

somebody1234 replied:
sounds good to me. i think our current format is just a convention, so the above is fine for me

somebody1234 replied:
although to be fair i don't think we technically need markdown. not that i'm _against_ it, but it does mean it's harder to split it onto two lines

somebody1234 replied:
yeah, actually i use TODO highlight too. and was a bit sad that the current format makes the highlighting not happen (with default settings, at least)

somebody1234 commented on app/gui/src/dashboard/services/remoteBackendPaths.ts:
```diff
@@ -204,8 +204,12 @@ export function getCheckoutSessionPath(checkoutSessionId: backend.CheckoutSessio
 /** The ID of the directory containing the home directories of all users. */
-export const USERS_DIRECTORY_ID = DirectoryId('directory-0000000000000000000000users')
+export const USERS_DIRECTORY_ID = backend.DirectoryId(
+  'directory-0000000000000000000000users' as const,
```
this is weird, is the `as const` actually necessary?

MrFlashAccount replied:
Good catch!

somebody1234 commented on app/gui/src/dashboard/services/RemoteBackend.ts:
```diff
@@ -48,17 +49,96 @@ function responseIsSuccessful(response: Response) {
+ * @param id - The user group ID.
+ * @returns The ID.
+ */
+export function extractIdFromUserGroupId(id: backend.UserGroupId) {
```
i think a bunch of these utility functions should be in `Backend.ts`, wdyt?

MrFlashAccount replied:
Maybe, but they seems unrelated to the LocalBackend?

somebody1234 commented on app/gui/src/dashboard/services/RemoteBackend.ts:
```diff
@@ -48,17 +49,96 @@ function responseIsSuccessful(response: Response) {
+ * Convert a user group ID to a directory ID.
+ */
+export function userGroupIdToDirectoryId(id: backend.UserGroupId): backend.DirectoryId {
+  return DirectoryId(`directory-${id.replace(/^usergroup-/, '')}` as const)
```
use `extractIdFromUserGroupId` here, similarly with other functions below

somebody1234 commented on app/gui/src/dashboard/providers/AuthProvider.tsx:
```diff
@@ -693,3 +698,10 @@ export function useFullUserSession(): FullUserSession {
+
+/** A React context hook returning the user session for a user that is fully logged in. */
+export function useUser() {
+  const { user } = useFullUserSession()
+
+  return user
```
(comment only, no action needed) this is truly one of the utility functions of all time

MrFlashAccount replied:
😁😁😁

Love this comment

somebody1234 commented on app/gui/src/dashboard/layouts/Drive/Categories/Category.ts:
```diff
@@ -0,0 +1,309 @@
+/** @file The categories available in the category switcher. */
```
weird, not sure why git didn't think this file was moved

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+}
+
+/**
+ * Individual item in the path.
```
incorrect docs

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+  const { getCategoryByDirectoryId } = useCloudCategoryList()
+
+  // Parents path is a string of directory ids separated by slashes.
+  const splittedPath = parentsPath.split('/').filter(isDirectoryId)
```
nit: technically the paste tense of `split` is `split`. this might be clearer though

MrFlashAccount replied:
Yeah, see: https://upskillstutor.com/news/past-tense-of-split/

Even though _technically_ `splitted` is correct, though uncommon. Will replace, thanks!

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+    const rootDirectoryInThePath = pathToDirectory.at(0)
+
+    // This should never happen, as we always have the root directory in the path.
+    // If it happens, it means you've skrewed up
```
not exactly professional, we can probably leave this line out

MrFlashAccount replied:
:D 

It certainly not professional, but I'm fine removing it.

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+
+    // If the target directory is null, we assume that this directory is outside of the current tree (in another category)
+    // Which is the default, because path displays in the recent and trash folders.
+    // But sometimes user might delete a directory with it's whole content, and in that case we'll find it in the tree
```
nit: `because the path`, `sometimes the user`, `it's` -> `its`, `because parent` -> `because the parent`

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+    const targetDirectoryNode = getAssetNodeById(targetDirectory)
+
+    if (targetDirectoryNode == null && rootDirectoryInThePath.categoryId != null) {
+      // we reassign the variable only to make ts happy here.
```
`We`, `TypeScript`

somebody1234 commented on app/gui/src/dashboard/components/dashboard/column/PathColumn.tsx:
```diff
@@ -0,0 +1,235 @@
+
+    // If the root category is not found it might mean
+    // that user is no longer have access to this root
+    // Usually this could happen if user was removed from the organization
```
`user` -> `the user` both here and on the line above. also optional but you could clarify what "this root" is. "this root directory"?

MrFlashAccount replied:
Yeah, you right here, updated the comment

somebody1234 commented on app/gui/src/dashboard/layouts/Drive/Categories/categoriesHooks.tsx:
```diff
@@ -0,0 +1,411 @@
+ * Hooks for working with categories.
+ * Categories are shortcuts to specific directories in the Cloud, e.g. team spaces, recent and trash
+ * It's not the same as the categories like LocalBackend
+ * TODO: Improve performance and add ability to subscribe to individual category values
```
ideally this TODO should be linked to an issue too

MrFlashAccount replied:
I think the todo here isn't strictly necessary as the perf impact is not measured yet.

somebody1234 reviewed (approved):
CR :white_check_mark:

MrFlashAccount committed:
Fix nits and comments

MrFlashAccount committed:
Fix comment

MrFlashAccount committed:
Remove await

MrFlashAccount committed:
Add issue for todo

MrFlashAccount committed:
Remove TODO

MrFlashAccount committed:
Merge branch 'develop' into wip/sergeigarin/drive-bugs

# Conflicts:
#	app/gui/src/dashboard/components/AriaComponents/Tooltip/Tooltip.tsx
#	app/gui/src/dashboard/components/dashboard/AssetRow.tsx
#	app/gui/src/dashboard/layouts/AssetsTable.tsx
#	app/gui/src/dashboard/layouts/TabBar.tsx

MrFlashAccount committed:
Fix types
